lib/osut/utils.rb

```diff
@@ -3334,56 +3396,88 @@ def pointWithinPolygon?(p0 = nil, s = [], entirely = false)
+    n1 = OpenStudio.getOutwardNormal(p1)
+    n2 = OpenStudio.getOutwardNormal(p2)
+    return false if n1.empty?
+    return false if n2.empty?
+
+    n1.get.dot(n2.get).abs > 0.99
+  end
+
+  ##
+  # Validates whether a polygon can be considered a valid 'roof' surface, as per
+  # ASHRAE 90.1 & Canadian NECBs, i.e. outward normal within 60° from vertical
+  #
+  # @param pts [Set<OpenStudio::Point3d>] OpenStudio 3D points
+  #
+  # @return [Bool] if considered a roof surface
+  # @return [false] if invalid input (see logs)
+  def roof?(pts = nil)
+    ray = OpenStudio::Point3d.new(0,0,1) - OpenStudio::Point3d.new(0,0,0)
+    dut = Math.cos(60 * Math::PI / 180)
+    pts = poly(pts, false, true, true)
+    return false if pts.empty?
 
-    pl1 = OpenStudio::Plane.new(p1)
-    pl2 = OpenStudio::Plane.new(p2)
+    dot = ray.dot(OpenStudio.getOutwardNormal(pts).get)
+    return false if dot.round(2) <= 0
+    return true  if dot.round(2) == 1
 
-    pl1.outwardNormal.dot(pl2.outwardNormal).abs > 0.99
+    dot.round(4) >= dut.round(4)
   end
 
   ##
-  # Validates whether a polygon faces upwards.
+  # Validates whether a polygon faces upwards, harmonized with OpenStudio
+  # Utilities' "alignZPrime" function.
   #
   # @param pts [Set<OpenStudio::Point3d>] OpenStudio 3D points
   #
   # @return [Bool] if facing upwards
   # @return [false] if invalid input (see logs)
   def facingUp?(pts = nil)
-    up  = OpenStudio::Point3d.new(0,0,1) - OpenStudio::Point3d.new(0,0,0)
-    pts = poly(pts, false, true, false)
-    return false if pts.empty?
-
-    pts = getNonCollinears(pts, 3)
+    ray = OpenStudio::Point3d.new(0,0,1) - OpenStudio::Point3d.new(0,0,0)
+    pts = poly(pts, false, true, true)
     return false if pts.empty?
 
-    OpenStudio::Plane.new(pts).outwardNormal.dot(up) > 0.99
+    OpenStudio.getOutwardNormal(pts).get.dot(ray) > 0.99
   end
 
   ##
-  # Validates whether a polygon faces downwards.
+  # Validates whether a polygon faces downwards, harmonized with OpenStudio
+  # Utilities' "alignZPrime" function.
   #
   # @param pts [Set<OpenStudio::Point3d>] OpenStudio 3D points
   #
   # @return [Bool] if facing downwards
   # @return [false] if invalid input (see logs)
   def facingDown?(pts = nil)
-    lo  = OpenStudio::Point3d.new(0,0,-1) - OpenStudio::Point3d.new(0,0,0)
-    pts = poly(pts, false, true, false)
+    ray = OpenStudio::Point3d.new(0,0,-1) - OpenStudio::Point3d.new(0,0,0)
+    pts = poly(pts, false, true, true)
     return false if pts.empty?
 
-    pts = getNonCollinears(pts, 3)
+    OpenStudio.getOutwardNormal(pts).get.dot(ray) > 0.99
+  end
+
+  ##
+  # Validates whether surface can be considered 'sloped' (i.e. not ~flat, as per
+  # OpenStudio Utilities' "alignZPrime"). A vertical polygon returns true.
+  #
+  # @param pts [Set<OpenStudio::Point3d>] OpenStudio 3D points
+  #
+  # @return [Bool] whether surface is sloped
+  # @return [false] if invalid input (see logs)
+  def sloped?(pts = nil)
```

Comment:
Tightly harmonizing with _OpenStudio_'s _alignZPrime_ rules on what is **flat** vs **sloped**. This had been an issue when allocating skylights to very low sloped roofs.